modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java

```diff
@@ -1225,6 +1232,82 @@ private void distributedExchange() throws IgniteCheckedException {
         initDone();
     }
 
+    /**
+     *
+     */
+    private void onDistributedExchangeTimeout() {
+        if (!isDone()) {
+            synchronized (mux) {
+                if (crd.isLocal()) {
+                    Set<UUID> remaining0 = new HashSet<>(this.remaining);
+
+                    log.warning("Exchange timeout reached on coordinator node, kicking pending nodes: " +
+                        S.arrayToString(UUID.class, remaining0.toArray()));
+
+                    for (UUID remainingUuid : remaining0)
+                        cctx.discovery().failNode(remainingUuid, null);
+                }
+                else {
+                    try {
+                        log.warning("Exchange timeout reached on non-coordinator node, " +
+                            "sending check request to coordinator");
+
+                        cctx.kernalContext().io().sendToGridTopic(
+                            crd, GridTopic.TOPIC_EXCHANGE, new PartitionsExchangeFinishedCheckRequest(), SYSTEM_POOL);
+
+                        cctx.time().schedule(() -> {
+                            if (!isDone())
+                                cctx.kernalContext().failure().process(
+                                    new FailureContext(FailureType.CRITICAL_ERROR,
+                                        new TimeoutException("Last exchange check response wasn't received.")));
```

Comment:
I asked 4 people to read this message, no one did understand what this means. I don't think that regular user would understand this message in logs. Could you please add more details here too?